Jetty 12.0.x documentation (#9096)
eclipse/jetty.project: documentation/jetty-documentation/src/main/asciidoc/operations-guide/begin/start.adoc

[[og-begin-start]]
==== Starting Jetty

Eclipse Jetty as a standalone server has no graphical user interface; configuring and running the server is done from the command line.

> First we need to create a `$JETTY_BASE` directory.

Reviewer: We -> You.

----
$ JETTY_BASE=/path/to/jetty.base
$ mkdir $JETTY_BASE
$ cd $JETTY_BASE
----

> We'll start Jetty by executing `$JETTY_HOME/start.jar` from this directory. However, if we try to start Jetty from an empty `$JETTY_BASE`, it will complain that we haven't enabled any modules:

Reviewer: We -> You.
Present tense.
New sentence on new line.

include::jetty[]
----

> Jetty uses a xref:og-modules[module] system to configure and assemble its web server; these modules are enabled and configured in xref:og-arch-jetty-base[`$JETTY_BASE`]. Since the `$JETTY_BASE` directory we just created is empty, Jetty has no configuration it can use to assemble the server.

Reviewer: ```suggestion
Jetty uses a xref:og-modules[module system] to configure and assemble the server; these modules are enabled and configured in xref:og-arch-jetty-base[`$JETTY_BASE`].
Since the `$JETTY_BASE` directory you just created is empty, Jetty has no configuration it can use to assemble the server.
```

NOTE: See the xref:og-arch[architecture section] of this document for more information on the design of Jetty's module system.

You can explore what modules are available with the `--list-modules` flag:

----
$ java -jar $JETTY_HOME/start.jar --list-modules=*
----

> Let's try to enable the xref:og-protocols-http[`http`] module.

Reviewer: Let's -> You (it was wrong even before) 😅

NOTE: If you want to enable support for protocols like secure HTTP/1.1 or HTTP/2 or HTTP/3, or want to configure Jetty behind a load balancer, read xref:og-protocols[this section].

----
$ java -jar $JETTY_HOME/start.jar --add-module=http
----

[source,subs=quotes,options=nowrap]
----
include::jetty[args="--add-module=http",highlight="([a-z\-]+ *transitively enabled)"]
----

> Note that Jetty has enabled a number of transitive dependencies (`server`, `logging-jetty`, and so on) in order to activate the `http` module.

Reviewer: ```suggestion
When Jetty enables the `http` module, it also automatically enables a number of transitive dependencies of the `http` module, such as the `server` module, the `logging-jetty` module, and so on.
```

> Now we can start Jetty:

Reviewer: We -> You.

include::jetty[args="--module=http",highlight="(\{.*:8080})"]
----

> Jetty is listening on port `8080` for clear-text HTTP/1.1 connections. However, since it has no web applications deployed, it will just reply with `404 Not Found` to every request.

Reviewer: New sentence on new line.

> Before we xref:og-begin-deploy[deploy our first web application], let's see what's happened to the `$JETTY_BASE` directory now that we've enabled the `http` module:

Reviewer: We/Our -> You/Your.

[source,subs=verbatim]
----

> $ tree $JETTY_BASE

Reviewer: Remove.

> /path/to/jetty.base

Reviewer: Same comment as previously.

├── resources
│   └── jetty-logging.properties <1>
└── start.d <2>
    └── http.ini <3>
----

> <1> The `resources/jetty-logging.properties` file configures the server's logging level. This file was auto-generated when the `jetty-logging` module was activated as a transitive dependency of the `http` module.

Reviewer: Use semicolon instead of period to keep a single sentence.

> <2> The `start.d/` directory contains configuration files for any modules you have explicitly activated.

Reviewer: ```suggestion
<2> The `start.d/` directory contains the `+*.ini+` configuration files for any modules you have explicitly activated.
```

<3> The `start.d/http.ini` file is the `http` module configuration file, where you can specify values for the xref:og-module-http[`http` module properties].

> NOTE: By default, Jetty does not generate configuration files in `start.d/` for modules activated as transitive dependencies. To manually configure such a module, you should activate it directly via Jetty's `--add_module` flag.

Reviewer: ```suggestion
[NOTE]
====
By default, Jetty does not generate `+*.ini+` configuration files in `start.d/` for modules activated as transitive dependencies. 
To manually configure such modules, you should activate them directly via Jetty's `--add-modules` flag.
====
```

> In the `http.ini` file we can find the following (among other contents):

Reviewer: We -> You.

----

<1> This line enables the `http` module and should not be modified.
<2> This commented line specifies the default value for the `jetty.http.port` property, which is the network port that Jetty uses to listen for clear-text HTTP connections.

> Let's change that default port. Open `http.ini`, uncomment the line containing `jetty.http.port=`, and change its value to `9999`:

Reviewer: Let's -> You.
New sentence on new line.

include::jetty[args="--module=http jetty.http.port=9999",highlight="(\{.*:9999})"]
----

> You can also specify the value of a module property when you start up Jetty. A property value specified on the command-line in this way will *override* the value configured in a module's `.ini` file.

Reviewer: New sentence on new line.
Use \`+*.ini+\` to indicate the `ini` files.
